bundles/org.openhab.binding.iammeter/src/main/java/org/openhab/binding/iammeter/internal/IammeterHandler.java

```diff
@@ -0,0 +1,208 @@
+                        State state = getDecimal(iammeterData.get(keyWord).getAsJsonArray().get(channelConfig.getRow())
+                                .getAsJsonArray().get(channelConfig.getCol()).toString());
+                        updateState(channel.getUID(), state);
+                    }
+                }
+            }
+            if (bRemoveChannels) {
+                if (!bExtraChannelRemoved) {
+                    thingStructureChanged(channelProfix);
+                }
+            }
+            stream.close();
+            updateStatus(ThingStatus.ONLINE);
+            // Very rudimentary Exception differentiation
+        } catch (IOException e) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                    "Communication error with the device: " + e.getMessage());
+        } catch (JsonSyntaxException je) {
+            logger.warn("Invalid JSON when refreshing source {}: {}", getThing().getUID(), je.getMessage());
+        }
+    }
+
+    protected void thingStructureChanged(String channelProfix) {
+        List<ChannelUID> noUsedItems = new ArrayList<>();
+        noUsedItems.add(new ChannelUID(channelProfix + ":frequency_a"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":pf_a"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":voltage_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":current_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":power_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":importenergy_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":exportgrid_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":frequency_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":pf_b"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":voltage_c"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":current_c"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":power_c"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":importenergy_c"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":exportgrid_c"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":frequency_c"));
+        noUsedItems.add(new ChannelUID(channelProfix + ":pf_c"));
+        ThingBuilder thingBuilder = editThing();
+        for (ChannelUID chl : noUsedItems) {
+            thingBuilder.withoutChannel(chl);
+        }
+        updateThing(thingBuilder.build());
+        bExtraChannelRemoved = true;
+    }
+
+    private State getDecimal(String value) {
+        try {
+            return QuantityType.valueOf(Float.parseFloat(value), SmartHomeUnits.VOLT);
+        } catch (NumberFormatException e) {
+            return UnDefType.UNDEF;
+        }
+    }
+
+    @Override
+    public void dispose() {
+        ScheduledFuture<?> refreshJob = this.refreshJob;
+        if (refreshJob != null && !refreshJob.isCancelled()) {
```

Comment:
No harm in canceling a task that is already cancelled.
```suggestion
        if (refreshJob != null) {
```

Reply:
removed.